docs/proposals/kubelet-eviction.md

```diff
@@ -241,11 +265,102 @@ only has guaranteed pod(s) remaining, then the node must choose to evict a
+1. The `root` filesystem that kubelet uses for volumes, daemon logs, etc.
+1. The `image` filesystem that container runtimes uses for storing images and container writable layers.
+`kubelet` does not care about any other filesystems.
+
+If the eviction threholds have been met for disk, kubelet will perform the following
+tasks to free up disk space in that order.
+
+1. Delete (or rotate) logs belonging to dead pods or previous incarnations of containers.
+2. Delete unused images on the nodes.
+3. Evict pods.
+
+Let's explore each option in detail.
+
+### Delete logs of dead pods/containers
+
+As of today, logs are tied to container's lifetime. `kubelet` keeps dead containers around,
+to provide access to logs.
+In the future, if we store logs of dead containers outside of the container itself, then
+kubelet can delete these logs to free up disk space.
+Until then, kubelet can delete dead containers to free up disk space. If images are stored in a separate `image`
+filesystem, kubelet will not attempt to delete containers, unless the `image` filesystem has met disk
+eviction thresholds
+
+Once the lifetime of containers and logs are split, kubelet can support more user friendly policies
+around log evictions. `kubelet` can delete logs of the oldest containers first.
+Since logs from the first and the most recent incarnation of a container is the most important for most applications,
+kubelet can try to preserve these logs and aggresively delete logs from other container incarnations.
+
+### Delete unused images
+
+`kubelet` performs image garbage collection based on thresholds today. It uses a high and a low watermark.
+Whenever disk usage exceeds the high watermark, it removes images until the low watermark is reached.
+`kubelet` employs a LRU policy when it comes to deleting images.
+
+If `image` and `root` filesystems are unified, then `kubelet` will attempt to delete images whenever the thresholds are met
+and there are no more old logs to be deleted.
+
+If the `image` filesystem is separate from the `root` filesystem, `kubelet` will track the usage of `image` filesystem,
+and delete images only when `image` filesystem usage exceeds.
+
+### Evict pods
+
+There is no ability to specify disk limits for pods/containers today.
+Disk is a best effort resource. If required, kubelet can evict pods one at a time.
+`kubelet` will follow the [Eviction Strategy](#eviction-strategy) mentioned above for making eviction decisions.
+
+If `root` filesystem has met eviction thresholds, and there is no dedicated `image` filesystem,
+`kubelet` will first attempt to delete old containers first, then delete images,
+and then eventually evict pods if required.
+
+If `root` filesystem has met eviction thresholds, and if there is a dedicated `image` filesystem,
+`kubelet` will attempt to evict pods directly.
+
+If the `image` filesystem has met eviction thresholds, `kubelet` will first attempt to delete old containers first,
```

Comment:
If there are a candidate set of dead containers to delete, can it rank dead containers using the priority described with the original flags:

```
maximum-dead-containers-per-container
minimum-container-ttl-duration
```

I agree we should always remove dead containers before evicting active pods, but it does seem like we could influence which dead containers to choose.  Maybe that is too complicated and disk alone is fair, but I suspect some folks will want a more nuanced policy here.